com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs

```diff
@@ -207,6 +209,104 @@ void SetPerCameraShaderVariables(CommandBuffer cmd, ref CameraData cameraData)
+            Vector3 cameraToWorldMatrixAxisY = new Vector3(cameraToWorldMatrix.m01, cameraToWorldMatrix.m11, cameraToWorldMatrix.m21);
+            Vector3 cameraToWorldMatrixAxisZ = new Vector3(cameraToWorldMatrix.m02, cameraToWorldMatrix.m12, cameraToWorldMatrix.m22);
+
+            Vector3 front = cameraToWorldMatrixAxisZ;
+
+            Vector3 worldUp = Vector3.up;
+            Vector3 cross = Vector3.Cross(front, worldUp);
+            billboardTangent = !Mathf.Approximately(cross.sqrMagnitude, 0.0f)
+                ? cross.normalized
+                : cameraToWorldMatrixAxisX;
+
+            billboardNormal = Vector3.Cross(worldUp, billboardTangent);
+            billboardNormal = !Mathf.Approximately(billboardNormal.sqrMagnitude, 0.0f)
+                ? billboardNormal.normalized
+                : cameraToWorldMatrixAxisY;
+
+            // SpeedTree generates billboards starting from looking towards X- and rotates counter clock-wisely
+            Vector3 worldRight = new Vector3(0, 0, 1);
+            // signed angle is calculated on X-Z plane
+            float s = worldRight.x * billboardTangent.z - worldRight.z * billboardTangent.x;
+            float c = worldRight.x * billboardTangent.x + worldRight.z * billboardTangent.z;
+            cameraXZAngle = Mathf.Atan2(s, c);
+
+            // convert to [0,2PI)
+            if (cameraXZAngle < 0)
+                cameraXZAngle += 2 * Mathf.PI;
+        }
+
+        // TODO: Move this to SetPerCameraShaderVariables once we get rid of context.SetupCameraProperties
+        private void SetPerCameraClippingPlaneProperties(CommandBuffer cmd, in CameraData cameraData)
+        {
+            // We need this to correctly handle projection inversion if needed
+            Matrix4x4 projectionMatrix = GL.GetGPUProjectionMatrix(cameraData.GetProjectionMatrix(), cameraData.targetTexture != null);
+
+            Matrix4x4 viewMatrix = cameraData.GetViewMatrix();
+
+            Matrix4x4 viewProj = CoreMatrixUtils.MultiplyProjectionMatrix(projectionMatrix, viewMatrix, cameraData.camera.orthographic);
+            Plane[] planes = new Plane[6];
+            GeometryUtility.CalculateFrustumPlanes(viewProj, planes);
+
+            static Vector4 PlaneToVector(Plane plane)
+            {
+                return new Vector4(plane.normal.x, plane.normal.y, plane.normal.z, plane.distance);
+            }
+
+            cmd.SetGlobalVector(ShaderPropertyId.cameraWorldClipPlanes0, PlaneToVector(planes[0]));
+            cmd.SetGlobalVector(ShaderPropertyId.cameraWorldClipPlanes1, PlaneToVector(planes[1]));
+            cmd.SetGlobalVector(ShaderPropertyId.cameraWorldClipPlanes2, PlaneToVector(planes[2]));
+            cmd.SetGlobalVector(ShaderPropertyId.cameraWorldClipPlanes3, PlaneToVector(planes[3]));
+            cmd.SetGlobalVector(ShaderPropertyId.cameraWorldClipPlanes4, PlaneToVector(planes[4]));
+            cmd.SetGlobalVector(ShaderPropertyId.cameraWorldClipPlanes5, PlaneToVector(planes[5]));
```

Comment:
It's weird to me we upload 6 individual constants instead of a vector array, is that due to built-in compatibility?

Reply:
This is from builtin, I think I can make it array, going to check